# Relatório de Análise V

## Tratamento de Dados Faltantes

In [0]:
import pandas as pd

In [0]:
dados = pd.read_csv('aluguel_residencial.csv', sep = ';')

In [0]:
dados.head(10)

Tipo           Bairro  Quartos  Vagas  Suites  Area    Valor  \
0           Quitinete       Copacabana        1      0       0    40   1700.0   
1                Casa  Jardim Botânico        2      0       1   100   7000.0   
2         Apartamento           Centro        1      0       0    15    800.0   
3         Apartamento     Higienópolis        1      0       0    48    800.0   
4         Apartamento     Vista Alegre        3      1       0    70   1200.0   
5         Apartamento         Cachambi        2      0       0    50   1300.0   
6  Casa de Condomínio  Barra da Tijuca        5      4       5   750  22000.0   
7  Casa de Condomínio            Ramos        2      2       0    65   1000.0   
8         Apartamento           Centro        1      0       0    36   1200.0   
9         Apartamento           Grajaú        2      1       0    70   1500.0   

   Condominio  IPTU  
0       500.0  60.0  
1         NaN   NaN  
2       390.0  20.0  
3       230.0   NaN  
4         NaN   NaN  
5       301.0  17.0  
6         NaN   NaN  
7         NaN   NaN  
8         NaN   NaN  
9       642.0  74.0

O primeiro deles é *isnull()*. Tal método irá gerar um DataFrame booleano, em que a observação marcada como **True caracteriza um dado nulo**, isto é, um missing.

In [4]:
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False        True   True
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False   True
4      False   False    False  False   False  False  False        True   True
5      False   False    False  False   False  False  False       False  False
6      False   False    False  False   False  False  False        True   True
7      False   False    False  False   False  False  False        True   True
8      False   False    False  False   False  False  False        True   True
9      False   False    False  False   False  False  False       False  False
10     False   False    False  False   False  False  False       False  False
11     False   False    False  False   False  False  False       False  False
12     False   False    False  False   False  False  False       False   True
13     False   False    False  False   False  False  False       False  False
14     False   False    False  False   False  False  False        True   True
15     False   False    False  False   False  False  False       False  False
16     False   False    False  False   False  False  False       False  False
17     False   False    False  False   False  False  False       False  False
18     False   False    False  False   False  False  False       False   True
19     False   False    False  False   False  False  False       False   True
20     False   False    False  False   False  False  False       False   True
21     False   False    False  False   False  False  False       False  False
22     False   False    False  False   False  False  False       False  False
23     False   False    False  False   False  False  False       False  False
24     False   False    False  False   False  False  False        True   True
25     False   False    False  False   False  False  False       False   True
26     False   False    False  False   False  False  False       False  False
27     False   False    False  False   False  False  False       False  False
28     False   False    False  False   False  False  False       False   True
29     False   False    False  False   False  False  False       False  False
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
22550  False   False    False  False   False  False  False       False  False
22551  False   False    False  False   False  False  False       False  False
22552  False   False    False  False   False  False  False       False  False
22553  False   False    False  False   False  False  False       False  False
22554  False   False    False  False   False  False  False       False  False
22555  False   False    False  False   False  False  False       False   True
22556  False   False    False  False   False  False  False       False   True
22557  False   False    False  False   False  False  False       False  False
22558  False   False    False  False   False  False  False        True  False
22559  False   False    False  False   False  False  False       False  False
22560  False   False    False  False   False  False  False       False  False
22561  False   False    False  False   False  False  False       False  False
22562  False   False    False  False   False  False  False       False  False
22563  False   False    False  False   False  False  False       False  False
22564  False   False    False  False   False  False  False       False  False
22565  False   False    False  False   False  False  False       False  False
22566  False   False    False  False   False  False  False       False  False
22567  False   False    False  False   False  False  False       False  False
22568  False   False    False  False   False  False  False       False   True
22569  False   False    False 

Temos, ainda, o método notnull(), que funciona exatamente da maneira inversa ao isnull(): se a informação for nula, será utilizada a notação False.

In [5]:
dados.notnull()

Tipo  Bairro  Quartos  Vagas  Suites  Area  Valor  Condominio   IPTU
0      True    True     True   True    True  True   True        True   True
1      True    True     True   True    True  True   True       False  False
2      True    True     True   True    True  True   True        True   True
3      True    True     True   True    True  True   True        True  False
4      True    True     True   True    True  True   True       False  False
5      True    True     True   True    True  True   True        True   True
6      True    True     True   True    True  True   True       False  False
7      True    True     True   True    True  True   True       False  False
8      True    True     True   True    True  True   True       False  False
9      True    True     True   True    True  True   True        True   True
10     True    True     True   True    True  True   True        True   True
11     True    True     True   True    True  True   True        True   True
12     True    True     True   True    True  True   True        True  False
13     True    True     True   True    True  True   True        True   True
14     True    True     True   True    True  True   True       False  False
15     True    True     True   True    True  True   True        True   True
16     True    True     True   True    True  True   True        True   True
17     True    True     True   True    True  True   True        True   True
18     True    True     True   True    True  True   True        True  False
19     True    True     True   True    True  True   True        True  False
20     True    True     True   True    True  True   True        True  False
21     True    True     True   True    True  True   True        True   True
22     True    True     True   True    True  True   True        True   True
23     True    True     True   True    True  True   True        True   True
24     True    True     True   True    True  True   True       False  False
25     True    True     True   True    True  True   True        True  False
26     True    True     True   True    True  True   True        True   True
27     True    True     True   True    True  True   True        True   True
28     True    True     True   True    True  True   True        True  False
29     True    True     True   True    True  True   True        True   True
...     ...     ...      ...    ...     ...   ...    ...         ...    ...
22550  True    True     True   True    True  True   True        True   True
22551  True    True     True   True    True  True   True        True   True
22552  True    True     True   True    True  True   True        True   True
22553  True    True     True   True    True  True   True        True   True
22554  True    True     True   True    True  True   True        True   True
22555  True    True     True   True    True  True   True        True  False
22556  True    True     True   True    True  True   True        True  False
22557  True    True     True   True    True  True   True        True   True
22558  True    True     True   True    True  True   True       False   True
22559  True    True     True   True    True  True   True        True   True
22560  True    True     True   True    True  True   True        True   True
22561  True    True     True   True    True  True   True        True   True
22562  True    True     True   True    True  True   True        True   True
22563  True    True     True   True    True  True   True        True   True
22564  True    True     True   True    True  True   True        True   True
22565  True    True     True   True    True  True   True        True   True
22566  True    True     True   True    True  True   True        True   True
22567  True    True     True   True    True  True   True        True   True
22568  True    True     True   True    True  True   True        True  False
22569  True    True     True   True    True  True   True        True   True
22570  True    True     True   True    True  True   Tru

Contudo, ainda que marquemos true ou false nos dados, ainda não temos a capacidade de analisá-los integralmente. Aprendemos logo nas primeiras aulas alguns usos do método info(), que nos exibe informações do DataFrame. 

In [6]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22580 entries, 0 to 22579
Data columns (total 9 columns):
Tipo          22580 non-null object
Bairro        22580 non-null object
Quartos       22580 non-null int64
Vagas         22580 non-null int64
Suites        22580 non-null int64
Area          22580 non-null int64
Valor         22571 non-null float64
Condominio    20765 non-null float64
IPTU          15795 non-null float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.6+ MB


Realizando uma seleção com base na Series que ele retorna para a coluna Valor usando o isnull()

In [7]:
dados[dados['Valor'].isnull()]

Tipo           Bairro  Quartos  ...  Valor  Condominio   IPTU
58    Apartamento  Barra da Tijuca        2  ...    NaN       970.0   68.0
1492  Apartamento             Leme        2  ...    NaN       878.0    NaN
1683         Casa     Campo Grande        3  ...    NaN         NaN    NaN
2012  Apartamento         Botafogo        2  ...    NaN      1010.0  170.0
2034  Apartamento       Copacabana        2  ...    NaN       850.0    NaN
4941         Casa     Campo Grande        3  ...    NaN         NaN    NaN
8568  Apartamento             Leme        2  ...    NaN       878.0    NaN
8947  Apartamento           Glória        3  ...    NaN       910.0  228.0
9149  Apartamento            Gávea        3  ...    NaN       880.0  221.0

[9 rows x 9 columns]

In [8]:
A = dados.shape[0]
dados.dropna(subset = ['Valor'], inplace = True)
B = dados.shape[0]
A - B

9

In [9]:
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

## Tratamento de Dados Faltantes (continuação)

verificaremos quantas assinaturas nulas temos em Condominio:

In [11]:
dados[dados['Condominio'].isnull()].shape[0]

1813

Agora precisamos filtrar os dados de acordo com o tipo Apartamento.

In [0]:
selecao = (dados['Tipo'] == 'Apartamento') & (dados['Condominio'].isnull())

Contudo, se nos atentarmos a nossa seleção, verificaremos que ela nos fornece exatamente as assinaturas que deverão ser removidas do banco de dados. Reaproveitaremos a estrutura já utilizada anteriormente neste notebook, e passaremos a seleção que criamos (selecao). Para eliminar as assinaturas nulas selecionadas ao invés de coletá-las, iremos adicionar ~ em selecao, que inverte a Series booleana.

In [13]:
A = dados.shape[0]
dados = dados[~selecao]
B = dados.shape[0]
A - B

745

Foram eliminadas 745 elementos da nossa base de dados. Vamos verificar quantas assinaturas nulas temos em Condominio agora:

In [0]:
dados[dados['Condominio'].isnull()].shape[0]

1068

Teremos o total de 1813 assinaturas nulas. Agora precisamos filtrar os dados de acordo do tipo Apartamento com Condomino nulo.

In [0]:
dados = dados.fillna({'Condominio': 0, 'IPTU': 0})

o resultado será 1068. O que faremos é manter esses dados, mas atribuir o valor 0 a eles. Temos uma função para esta ação fillna(), que receberá o valor 0 e inplance = True.

In [0]:
dados[dados['Condominio'].isnull()].shape[0]

0

Outra maneira de resolver a questão, é criar um dicionário que indicará os valores das variáveis. Podemos, inclusive, inserir valores diferentes para ambas as variáveis.

In [0]:
dados[dados['IPTU'].isnull()].shape[0]

0

Ao verificarmos novamente a quantidade de nulos existentes em nosso DataFrame, tanto para variável Condominio quanto para IPTU, o resultado será 0. Podemos verificar com mais clareza escrevendo:

In [0]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21826 entries, 0 to 22579
Data columns (total 9 columns):
Tipo          21826 non-null object
Bairro        21826 non-null object
Quartos       21826 non-null int64
Vagas         21826 non-null int64
Suites        21826 non-null int64
Area          21826 non-null int64
Valor         21826 non-null float64
Condominio    21826 non-null float64
IPTU          21826 non-null float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


Temos, portanto, 21826 registros e nenhum deles é nulo. Para finalizarmos essa etapa, devemos salvar este arquivo, de forma que o DataFrame seja atualizado com as modificações que realizamos. Para tanto, escreveremos - lembrando que não queremos incluir o índice:

In [0]:
dados.to_csv('dados/aluguel_residencial.csv', sep = ';', index = False)